external-crates/move/tools/move-package/src/resolution/dependency_graph.rs

```diff
@@ -141,83 +130,676 @@ struct DependencyTOML<'a>(PM::PackageName, &'a Dependency);
+                                 conflicting dependencies found for '{}':\n{}{}",
+                            root_package,
+                            sub_pkg_name,
+                            format_pkgs("\nNot found:", combined_pkgs),
+                            format_pkgs("\nNew:", sub_pkgs),
+                        )
+                    };
+                    bail!(msg);
+                }
+            } else {
+                // package being inserted different from existing package
+                if DependencyGraph::get_dep_override(
+                    root_package,
+                    &sub_pkg_name,
+                    overrides,
+                    dev_overrides,
+                    mode == DependencyMode::DevOnly,
+                )?
+                .is_none()
+                {
+                    // no override exists
+                    bail!(
+                        "When resolving dependencies for package {0}, \
+                         conflicting dependencies found:\n{1} = {2}\n{1} = {3}",
+                        root_package,
+                        sub_pkg_name,
+                        PackageWithResolverTOML(existing_pkg),
+                        PackageWithResolverTOML(&sub_pkg),
+                    );
+                }
+                // otherwise override (already inserted before graph combining even started) for a
+                // given package exists and all is good
+            }
+            true
+        } else {
+            // a package with a given name does not exist in the graph yet
+            self.package_table.insert(sub_pkg_name, sub_pkg.clone());
+            false
+        };
+        Ok(pkg_exists)
+    }
 
-        Ok(graph)
+    /// Helper function to get overrides for "regular" dependencies (`dev_only` is false) or "dev"
+    /// dependencies (`dev_only` is true).
+    fn get_dep_override<'a>(
+        root_pkg_name: PM::PackageName,
+        pkg_name: &PM::PackageName,
+        overrides: &'a BTreeMap<Symbol, Package>,
+        dev_overrides: &'a BTreeMap<Symbol, Package>,
+        dev_only: bool,
+    ) -> Result<Option<&'a Package>> {
+        // for "regular" dependencies override can come only from "regular" dependencies section,
+        // but for "dev" dependencies override can come from "regular" or "dev" dependencies section
+        if let Some(pkg) = overrides.get(pkg_name) {
+            // "regular" dependencies section case
+            if dev_overrides.get(pkg_name).is_some() {
+                bail!(
+                    "Conflicting \"regular\" and \"dev\" overrides of {} in {}",
+                    pkg_name,
+                    root_pkg_name
+                );
+            }
+            return Ok(Some(pkg));
+        } else if dev_only {
+            // "dev" dependencies section case
+            if let Some(dev_pkg) = dev_overrides.get(pkg_name) {
+                return Ok(Some(dev_pkg));
+            }
+        }
```

Comment:
nit: `bool::then_some` might help here.

```suggestion
        } else if let Some(dev_pkg) = dev_overrides.get(pkg_name) {
            // "dev" dependencies section case
            return Ok(dev_only.then_some(dev_pkg));
        }
```

Reply:
Clever!!!